### ISSUE TITLE
Overhaul CMakeLists.txt

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#4514]
I recently read a great tutorial on CMake for packaging: https://alexreinking.com/blog/building-a-dual-shared-and-static-library-with-cmake.html. Lots of good pointers on how to modernize a CMake setup. I'd like to take a stab at that over the next weeks. And clean some of the mess we patched over the years.

**Comment by kevinkreiser** [comment]
yeah this is the primary reason I stopped doing the ppa. it was non trivial to do static and dynamic in the same build whereas with autotools its was the default

**Comment by nilsnolde** [comment]
another thing I'd like to add to this, since it will become a topic soon:

so far we have been fairly relaxed about min version of cmake. I've been reading up about it a bit (e.g. [this article](https://alexreinking.com/blog/how-to-use-cmake-without-the-agonizing-pain-part-1.html)). and I think we should require whatever version is best for us to maintain this project. cmake makes it _crazy_ easy to update a version, even tarballs from their homepage are installable without root access. I don't think there's a good argument to make our own life harder just so some devs don't have to make sure their software is up to date. for debian/ubuntu there's kitware PPA's, other platforms are almost always on newest version anyways. and centos/rhel users.. well, they can build from source if need be.

once things are packaged and available over package managers, there's hardly any non-techies having to deal with this anyways.
